### PR TITLE
(rebased) New CMake targets-based build configuration

### PR DESCRIPTION
**Pull request by jlblancoc**
Just a follow-up PR after closing and rebasing all the commits in #27 .

### CODE REVIEW

**Comment by jlblancoc** [comment]
Status of this one: all tests pass in Travis (except one that times out). 
Will check on Windows too, but for that I'm preparing a separate PR to fix other MSVC-specific errors first...

**Comment by chrisbeall** [comment]
Taking a final look now! Meanwhile, please fix the new conflict?

**Comment by jlblancoc** [comment]
> Taking a final look now! Meanwhile, please fix the new conflict?

Great! Was doing it right now. It's fixed now (rebased and push'ed --force to my fork).

File: cmake/GtsamBuildTypes.cmake [hunk 1]
```diff
@@ -1,7 +1,45 @@
...
+# function:  append_config_if_not_empty(TARGET_VARIABLE build_type)
+# Auxiliary function used to merge configuration-specific flags into the
+# global variables that will actually be send to cmake targets.
+# -----------------------------------------------------------
+function(append_config_if_not_empty TARGET_VARIABLE_ build_type)
+  string(TOUPPER "${build_type}" build_type_toupper)
+  set(flags_variable_name "${TARGET_VARIABLE_}_${build_type_toupper}")
+  set(flags_ ${${flags_variable_name}})
+  if (NOT "${flags_}" STREQUAL "")
+    if ("${build_type_toupper}" STREQUAL "COMMON")
```
[the remark this answers is not in the document]
Comment (jlblancoc) [reply cmake/GtsamBuildTypes.cmake]: Had to use a more modern version of cmake to see that warning, thanks!

It's fixed now too.

**Comment by jlblancoc** [comment]
@chrisbeall , @dellaert : feel free of merging. Thanks.

**Comment by dellaert** [comment]
Excellent, I will merge.

**Comment by fishyai** [comment]
Hey y'all, I'm finding that this commit breaks the flag this PR seems to breaks the GTSAM_INSTALL_MATLAB_TOOLBOX flag on Ubuntu 16.04. Specifically, it was the commit titled "Refactor build flags via CMake target properties" that was included in this merge.

I get this error:

 CMake Error at cmake/GtsamMatlabWrap.cmake:106 (message):
   You must call find_package(GTSAM) before using wrap
 Call Stack (most recent call first):
   cmake/GtsamMatlabWrap.cmake:59 (wrap_library_internal)
   gtsam/CMakeLists.txt:214 (wrap_and_install_library)

I'm trying to figure out what exactly causes the error. If I figure it out, I'll send a PR.

**Comment by fishyai** [comment]
Ah I should add that this was with Cmake 3.14

**Comment by fishyai** [comment]
Here is the PR to fix the bug: https://github.com/borglab/gtsam/pull/89

**Pull request merged**